pkg/runtime/wfengine/wfengine_test.go

```diff
@@ -517,6 +517,104 @@ func TestRecreateRunningWorkflowFails(t *testing.T) {
 	}
 }
 
+// TestRetryWorkflowOnTimeout verifies that workflow operations are retried when they fail to complete.
+func TestRetryWorkflowOnTimeout(t *testing.T) {
+	const expectedCallCount = 3
+	actualCallCount := 0
+
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("FlakyWorkflow", func(ctx *task.OrchestrationContext) (any, error) {
+		// update this global counter each time the workflow gets invoked
+		actualCallCount++
+		if actualCallCount < expectedCallCount {
+			// simulate a hang for the first two calls
+			time.Sleep(5 * time.Minute)
+		}
+		return actualCallCount, nil
+	})
+
+	ctx := context.Background()
+	client, engine := startEngine(ctx, r)
+
+	// Set a really short timeout to override the default workflow timeout so that we can exercise the timeout
+	// handling codepath in a short period of time.
+	engine.SetWorkflowTimeout(1 * time.Second)
+
+	// Set a really short reminder interval to retry workflows immediately after they time out.
+	engine.SetActorReminderInterval(1 * time.Millisecond)
+
+	for _, opt := range GetTestOptions() {
+		t.Run(opt(engine), func(t *testing.T) {
+			actualCallCount = 0
+
+			id, err := client.ScheduleNewOrchestration(ctx, "FlakyWorkflow")
+			if assert.NoError(t, err) {
+				// Add a 5 second timeout so that the test doesn't take forever if something isn't working
+				timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+				defer cancel()
+
+				metadata, err := client.WaitForOrchestrationCompletion(timeoutCtx, id)
+				if assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, fmt.Sprintf("%d", expectedCallCount), metadata.SerializedOutput)
+				}
+			}
+		})
+	}
+}
+
+// TestRetryActivityOnTimeout verifies that activity operations are retried when they fail to complete.
+func TestRetryActivityOnTimeout(t *testing.T) {
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("FlakyWorkflow", func(ctx *task.OrchestrationContext) (any, error) {
+		var output int
+		err := ctx.CallActivity("FlakyActivity", nil).Await(&output)
+		return output, err
+	})
+
+	const expectedCallCount = 3
+	actualCallCount := 0
+
+	r.AddActivityN("FlakyActivity", func(ctx task.ActivityContext) (any, error) {
+		// update this global counter each time the activity gets invoked
+		actualCallCount++
+		if actualCallCount < expectedCallCount {
+			// simulate a hang for the first two calls
+			time.Sleep(5 * time.Minute)
+		}
+		return actualCallCount, nil
+	})
+
+	ctx := context.Background()
+	client, engine := startEngine(ctx, r)
+
+	// Set a really short timeout to override the default activity timeout (1 hour at the time of writing)
+	// so that we can exercise the timeout handling codepath in a short period of time.
+	engine.SetActivityTimeout(1 * time.Second)
```

Comment:
Is there a way we can use a mock clock in unit tests? We've seen many times that using the real wall clock leads to flakiness. If you can add support for mocking the time object, you can use the `clock` library then (it's already in use by some packages)

Reply:
I'm not familiar with the `clock` library or with mocking time more generally. I'll need to spend some time investigating how to do this.